apm-agent-core/src/main/java/co/elastic/apm/agent/bci/HelperClassManager.java

```diff
@@ -271,7 +290,148 @@ private synchronized T loadAndReferenceHelper(Class<?> classOfTargetClassLoader)
+        }
+
+        /**
+         * Gets all {@link RegisterMethodHandle} annotated methods and registers them in {@link MethodHandleDispatcher}
+         */
+        private static void registerMethodHandles(List<String> classesToInject, ClassLoader helperCL, @Nullable ClassLoader targetClassLoader, @Nullable ProtectionDomain protectionDomain) throws Exception {
+            ensureDispatcherForClassLoaderCreated(targetClassLoader, protectionDomain);
+            ConcurrentMap<String, MethodHandle> dispatcherForClassLoader = MethodHandleDispatcher.getDispatcherForClassLoader(targetClassLoader);
+            ConcurrentMap<String, Method> reflectionDispatcherForClassLoader = MethodHandleDispatcher.getReflectionDispatcherForClassLoader(targetClassLoader);
+
+            Class<MethodHandleRegisterer> methodHandleRegistererClass = (Class<MethodHandleRegisterer>) Class.forName(MethodHandleRegisterer.class.getName(), true, helperCL);
+            TypePool typePool = TypePool.Default.of(getAgentClassFileLocator());
+            for (String name : classesToInject) {
+                // check with Byte Buddy matchers, acting on the byte code as opposed to reflection first
+                // to avoid NoClassDefFoundErrors when the classes refer to optional types
+                boolean isHelperClass = !typePool.describe(name)
+                    .resolve()
+                    .getDeclaredMethods()
+                    .filter(isAnnotatedWith(named(RegisterMethodHandle.class.getName())))
+                    .isEmpty();
+                if (isHelperClass) {
+
+                    Class<?> clazz = Class.forName(name, true, helperCL);
+                    if (clazz.getClassLoader() != helperCL) {
+                        throw new IllegalStateException("Helper classes not loaded from helper class loader, instead loaded from " + clazz.getClassLoader());
+                    }
+                    try {
+                        // call in from the context of the created helper class loader so that helperClass.getMethods() doesn't throw NoClassDefFoundError
+                        Method registerStaticMethodHandles = methodHandleRegistererClass.getMethod("registerStaticMethodHandles", ConcurrentMap.class, ConcurrentMap.class, Class.class);
+                        registerStaticMethodHandles.invoke(null, dispatcherForClassLoader, reflectionDispatcherForClassLoader, clazz);
+                    } catch (Exception e) {
+                        logger.error("Exception while trying to register method handles for {}", name);
+                        if (e instanceof InvocationTargetException) {
+                            Throwable targetException = ((InvocationTargetException) e).getTargetException();
+                            if (targetException instanceof Exception) {
+                                throw (Exception) targetException;
+                            }
+                        }
+                        throw e;
+                    }
```

Comment:
Lazily injecting the helper classes has the downside that we wouldn't be able to leverage an upcoming feature in Byte Buddy. This feature would inject an `INVOKEDYNAMIC` instruction to call the helper method: https://github.com/raphw/byte-buddy/issues/830#issuecomment-603955721
This avoids the overhead of having to look up the method handle from a map.

However, there are a few things we can do to make this safer:
* We already only load classes where we checked with Byte Buddy matchers that they contain `@RegisterMethodHandle`-annotated methods.
* Call `getMethods` instead of `getDeclaredMethods`: avoids resolving types of all parameters of non-public methods. So we could make methods with problematic signatures private or package-private.
* Using the upcoming `bootstrap` method, I think we'd be able to get the method handles without having to use reflection. We'd need to do some testing but I suspect we'll need a fallback for Java 7 that does not rely on `INVOKEDYNAMIC`.

Reply:
> Lazily injecting the helper classes has the downside that we wouldn't be able to leverage an upcoming feature in Byte Buddy

Is it something we should rely on? And won't we have to do some adjustments anyhow?

> Call getMethods instead of getDeclaredMethods: avoids resolving types of all parameters of non-public methods. So we could make methods with problematic signatures private or package-private.

The nature of these problems is that they are very setup-dependant, so there are not "problematic signatures" per se. In most cases it will work, but somewhere (eg some module-dependency systems) these side effects may cause a lookup of a type that is not yet available.

Bottom line- it is safer and if the effort is not big, I think it worth it (as the effort to debug even a single related problem could be bigger). However, if the effort is substantial, or if you think it will require a rework effort in the foreseeable future - let's merge as is and rethink later if needed. It's really your call, either way is fine for me.